### ISSUE TITLE
Phase retrieval workflow

### ISSUE THREAD

**Issue by mb1069** [david-hoffman/pyOTF#18]
Hi David,
As with  #15 I am trying to model experimental PSFs to simulate equivalent training data,
I had however given up on using pyOTF after struggling to find reasonable results.

Since seeing the other issue, i've added calls to `center_data()` and `prep_data_for_PR` to my workflow, and have gotten significantly improved results. However, I was wondering if this could be taken further; are there any other steps I might be missing?

I have produced a 'minimal' example below, and included a gzip of the PSF;
```python
from pyotf import otf
from pyotf.otf import HanserPSF, apply_aberration
from pyotf.phaseretrieval import retrieve_phase
from pyotf.utils import prep_data_for_PR, center_data
from tifffile import imread, imshow
import numpy as np
import matplotlib.pyplot as plt


def mse(y, y_pred):
    return np.mean((y - y_pred) ** 2).sum()


model_kwargs = dict(
    wl=660,
    na=1.3,
    ni=1.51,
    res=85,
    size=32,
    zsize=41,
    zres=100,
    vec_corr="none",
    condition="none",
)

target_psf = imread('./2.tif')

target_psf = center_data(target_psf)

# Normalise PSF and cast to uint8
target_psf = target_psf / target_psf.max()
target_psf *= 255
target_psf = target_psf.astype(np.uint8)

target_psf = prep_data_for_PR(target_psf, multiplier=1.01)

# Retrieve phase for experimental PSF
PR_result = retrieve_phase(
    target_psf, model_kwargs, max_iters=1000, pupil_tol=0, mse_tol=0,
    phase_only=False,
)

PR_result.fit_to_zernikes(32)
PR_result.plot()
PR_result.plot_convergence()

# Simulate HanserPSF with parameters

result_psf = HanserPSF(**model_kwargs)
result_psf = apply_aberration(result_psf, PR_result.zd_result.mcoefs, PR_result.zd_result.pcoefs)

# Render side-by-side axial slices

result_psf = result_psf.PSFi

target_psf = target_psf / target_psf.max()
result_psf = result_psf / result_psf.max()
print(mse(target_psf, result_psf))
comb_psf = np.concatenate((target_psf, result_psf), axis=2)
comb_psf = comb_psf[slice(5, 40, 3)]
comb_psf = np.concatenate(comb_psf, axis=0)
imshow(comb_psf)
plt.show()
```
[2.tif.gz](https://github.com/david-hoffman/pyOTF/files/6028494/2.tif.gz)

Thanks again for developing this package; I would be happy to help describe the final workflow for other users in more detail.

**Comment by david-hoffman** [comment]
Can you attach the output images from your plot calls?

Which version are you using? I updated `prep_data_for_PR` to clip values at `0` here 5fb715337b6c85e4554d006cef86f7556bf0b18b.

Is the PSF data raw microscope data? If so I think it'd be better to leave it as is, or at least cast to a deeper bit-depth `int`.

One other thing, there are these two methods of the `PhaseRetrievalResult` object:
https://github.com/david-hoffman/pyOTF/blob/6046ef65ba1459143be3fea21bc75a53fa641e1b/pyotf/phaseretrieval.py#L223-L229

You might want to add a `PR_result.zd_result.plot()` to see how good the Zernike decomposition is.

This is connected to #13.

**Comment by mb1069** [comment]
Sure, 
Here is an example of the target_psf (left), difference between psfs (centre) and reconstructed psf (right)
![image](https://user-images.githubusercontent.com/9156542/108884656-23a81a00-7607-11eb-9e77-b7a5b741b33d.png)


Pcoef results for all points in the experimental image
![image](https://user-images.githubusercontent.com/9156542/108884398-d9269d80-7606-11eb-9038-04af3a4073b8.png)

Mcoef results
![image](https://user-images.githubusercontent.com/9156542/108884506-f8bdc600-7606-11eb-9767-97a31a793da9.png)


Thank you for the suggestions, I will give those a go

**Comment by david-hoffman** [comment]
Could you post the results of:
- `PR_result.plot()`
- `PR_result.zd_result.plot()`
- `PR_result.plot_convergence()`

Also, trying running your script changing the number of fitted Zernike Polynomials to the max supported, i.e. `PR_result.fit_to_zernikes(120)` and show your comparison plot.

To clarify, the bar plots you're showing at the end are the result of performing phase retrieval on multiple point sources in one or more fields of view?

**Comment by mb1069** [comment]
I upped the number of zerns to 120;
Pr_result.plt() and PR_result.zd_result.plot() are indistinguishable;
![image](https://user-images.githubusercontent.com/9156542/108887494-9ff02c80-760a-11eb-9e5f-00436f5f6a81.png)

plot_convergence():
![image](https://user-images.githubusercontent.com/9156542/108887670-d4fc7f00-760a-11eb-9589-893d9d1f502d.png)

The bar plots are from a single FOV, I believe. We're hoping to eventually customise the aberration modelling for different sections of the FOV but that's still a while away.

**Comment by david-hoffman** [comment]
Can you add the comb plot after reconstructing with more Zernikes?

**Comment by david-hoffman** [comment]
So you’re finding multiple emitters in a single FOV and doing phase retrieval on all of them, correct?

**Comment by mb1069** [comment]
Yup, the non-bar plots are a single example of reconstruction, and the bar plots represent results across all emitters.

Here's the 120-zern version of the comb plot
![image](https://user-images.githubusercontent.com/9156542/108893764-c1084b80-7611-11eb-821e-e5b4beaf6445.png)

**Comment by david-hoffman** [comment]
Ok here's my version:

```python
import pathlib

import matplotlib.pyplot as plt
import numpy as np
from pyotf.phaseretrieval import retrieve_phase
from pyotf.utils import prep_data_for_PR
from tifffile import imread


def mse(y, y_pred):
    return np.mean((y - y_pred) ** 2).sum()


model_kwargs = dict(
    wl=660,
    na=1.3,
    ni=1.51,
    res=85,
    size=32,
    zsize=41,
    zres=100,
    vec_corr="none",
    condition="none",
)

target_psf = imread(pathlib.Path(__file__).parent / "2.tif")

target_psf = prep_data_for_PR(target_psf, multiplier=1.0)

# Retrieve phase for experimental PSF
PR_result = retrieve_phase(target_psf, model_kwargs)

PR_result.fit_to_zernikes(120)
PR_result.plot()
PR_result.zd_result.plot()
PR_result.plot_convergence()

# Simulate HanserPSF with parameters

result_psf = PR_result.generate_zd_psf(sphase=slice(None))

# this part is very kludgy
PR_result.model.PSFi = target_psf
PR_result.model.plot_psf()

PR_result.model.PSFi = result_psf
PR_result.model.plot_psf()

target_psf = target_psf / target_psf.max()
result_psf = result_psf / result_psf.max()

print(mse(target_psf, result_psf))

plt.show()
```

There's a few things:
1. The algorithm is super sensitive to background, so you have to play with the `multiplier` keyword for `prep_data_for_PR`.
2. You shouldn't use `center_data` except when cropping (I fixed this in 151aa9ee51dcc67406d72c4e46787e47610b77d7).
3. A lot of the information about higher order aberrations is contained in the out-of-focus planes so you need to
  a.  Use a larger cropping box (assuming that the beads are far enough apart)
  b. Take data with higher SNR.

Here's the two plots:
<img width="868" alt="image" src="https://user-images.githubusercontent.com/3699983/108898283-e1adb280-75cb-11eb-805d-c8a3a5e47390.png">
<img width="868" alt="image" src="https://user-images.githubusercontent.com/3699983/108898303-e7a39380-75cb-11eb-94a0-9d83ab32fffc.png">

The top is the real data and the bottom is the reconstruction.

Note that you capture the in focus parts of the PSF accurately but not the out of focus parts due both to low SNR and too small of a cropping box.
